### PR TITLE
feat(pattern): add glob exclusion support

### PR DESCRIPTION
**Pull request by Osirisxxl**

### CODE REVIEW

File: lib/sw-precache.js [hunk 1]
```diff
@@ -157,9 +177,21 @@ function generate(params, callback) {
...
     params.stripPrefixMulti[params.stripPrefix] = params.replacePrefix;
 
+    var includedFilePathsArray = [];
+    var excludedFilePathsArray = [];
     params.staticFileGlobs.forEach(function(globPattern) {
-      var filesAndSizesAndHashes = getFilesAndSizesAndHashesForGlobPattern(
-        globPattern, params.outputFilePath);
+      var toPush = includedFilePathsArray;
+      var safePattern = globPattern.replace(path.sep, '/');
+      if (safePattern.charAt(0) === '!' && !(safePattern.charAt(1) === '(')) {
```
Comment (Flavien-Pensato) [inline lib/sw-precache.js]: safePattern.charAt(1) !== '(' 💡

Comment (Osirisxxl) [reply lib/sw-precache.js]: missed that ^^

**Review by Flavien-Pensato — APPROVED** [review]

File: lib/sw-precache.js [hunk 2]
```diff
@@ -157,9 +177,21 @@ function generate(params, callback) {
...
+      var toPush = includedFilePathsArray;
+      var safePattern = globPattern.replace(path.sep, '/');
+      if (safePattern.charAt(0) === '!' && safePattern.charAt(1) !== '(') {
+        toPush = excludedFilePathsArray;
+        safePattern = safePattern.substring(1);
+      }
+      toPush.push(glob.sync(safePattern));
+    });
+
+    includedFilePathsArray.forEach(function(filePaths) {
```
Comment (DevSide) [inline lib/sw-precache.js]: need to deduplicate exclude and include paths

File: lib/sw-precache.js [hunk 3]
```diff
@@ -51,13 +51,33 @@ function getFileAndSizeAndHashForFile(file) {
...
-  return glob.sync(globPattern.replace(path.sep, '/')).map(function(file) {
+/**
+ * Check if a filePath is within a list of filePaths
+ * @param {String} filePath
+ * @param {Array} filePathsArray
+ * @return {boolean}
+ */
+function match(filePath, filePathsArray) {
+  return filePathsArray.some(function(filePaths) {
+    return filePaths.some(function(otherFilePath) {
```
Comment (DevSide) [inline lib/sw-precache.js]: `excludedFilePathsArray` could be flatten to avoid a 2-dimensions search here

File: test/test.js [hunk 4]
```diff
@@ -101,6 +101,28 @@ describe('sw-precache core functionality', function() {
     });
   });
 
+
+  it('should ignore folder two', function(done) {
+    var configOne = {
+      logger: NOOP,
+      staticFileGlobs: ['test/data/**']
```
Comment (Osirisxxl) [inline test/test.js]: I'd rather test an equality with `configOne` targeting folder one 💡:
```js
var configOne = { 
  logger: NOOP,
  staticFileGlobs: ['test/data/one/**']
}
var configTwo = {
  logger: NOOP,
  staticFileGlobs: ['test/data/**', '!test/data/two/**']
}

// assert.equal(responseStringOne, responseStringTwo);
```
This seems more targeted and robust to me 🙂

Comment (Flavien-Pensato) [reply test/test.js]: Oh yeah good idea

**Review by fdubost — APPROVED** [review]

**Review by j-allison — APPROVED** [review]

**Pull request merged**